File: src/NodeParser/UndefinedTypeNodeParser.ts
import ts from "typescript";
import { Context } from "../NodeParser.js";
import { SubNodeParser } from "../SubNodeParser.js";
import { BaseType } from "../Type/BaseType.js";
import { UndefinedType } from "../Type/UndefinedType.js";

export class UndefinedTypeNodeParser implements SubNodeParser {
    public supportsNode(node: ts.KeywordTypeNode): boolean {
        return node.kind === ts.SyntaxKind.UndefinedKeyword;
    }
    public createType(node: ts.KeywordTypeNode, context: Context): BaseType {
        return new UndefinedType();
    }
}

File: internal/parser/undefined_type_node_parser.go
package parser

import (
	"github.com/microsoft/typescript-go/shim/ast"

	"github.com/vega/ts-json-schema-generator/internal/types"
)

// UndefinedTypeNodeParser handles the `undefined` keyword
// (src/NodeParser/UndefinedTypeNodeParser.ts).
type UndefinedTypeNodeParser struct{}

func NewUndefinedTypeNodeParser() *UndefinedTypeNodeParser {
	return &UndefinedTypeNodeParser{}
}

func (p *UndefinedTypeNodeParser) SupportsNode(node *ast.Node) bool {
	return node.Kind == ast.KindUndefinedKeyword
}

func (p *UndefinedTypeNodeParser) CreateType(_ *ast.Node, _ *Context, _ *types.ReferenceType) types.Type {
	return &types.UndefinedType{}
}
